cmd/kubeadm/app/phases/addons/dns/dns.go

```diff
@@ -398,29 +401,47 @@ var (
 )
 
 func isCoreDNSVersionSupported(client clientset.Interface) (bool, error) {
-	isValidVersion := true
-	coreDNSPodList, err := client.CoreV1().Pods(metav1.NamespaceSystem).List(
-		context.TODO(),
-		metav1.ListOptions{
-			LabelSelector: "k8s-app=kube-dns",
-		},
-	)
-	if err != nil {
-		return false, errors.Wrap(err, "unable to list CoreDNS pods")
+	var lastError error
+	var pods []v1.Pod
+	err := wait.PollImmediate(constants.APICallRetryInterval, 10 * time.Second, func() (bool, error) {
+		coreDNSPodList, err := client.CoreV1().Pods(metav1.NamespaceSystem).List(
+			context.TODO(),
+			metav1.ListOptions{
+				LabelSelector: "k8s-app=kube-dns",
+				FieldSelector: "status.phase=Running",
+			},
+		)
+		if err != nil {
+			lastError = err
+			return false, nil
+		}
+
+		if len(coreDNSPodList.Items) <= 0 {
+			lastError = errors.New("unable to list CoreDNS Running pods")
+			return false, nil
+		}
+		pods = coreDNSPodList.Items
+		return true, nil
+	})
+
+	if err != nil || len(pods) <= 0 {
+		return false, fmt.Errorf("%v: %v", err, lastError)
 	}
 
-	for _, pod := range coreDNSPodList.Items {
+	isValidVersion := true
+	for _, pod := range pods {
 		imageID := imageDigestMatcher.FindStringSubmatch(pod.Status.ContainerStatuses[0].ImageID)
 		if len(imageID) != 2 {
-			return false, errors.Errorf("unable to match SHA256 digest ID in %q", pod.Status.ContainerStatuses[0].ImageID)
+			lastError = errors.Errorf("pod %s unable to match SHA256 digest ID in %q", pod.GetName() ,pod.Status.ContainerStatuses[0].ImageID)
+			return false, lastError
 		}
 		// The actual digest should be at imageID[1]
 		if !migration.Released(imageID[1]) {
 			isValidVersion = false
```

Comment:
i don't think isValidVersion and lastError are needed.

here the function can return:
`fasle, errors.Errorf("unknown digest %q for pod %s", imageID[1], pod.GetName())`